File: applicatie/src/main/java/com/vdzon/administratie/bestelling/BestellingService.java
package com.vdzon.administratie.bestelling;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.vdzon.administratie.auth.SessionHelper;
import com.vdzon.administratie.crud.UserCrud;
import com.vdzon.administratie.dto.BestellingDto;
import com.vdzon.administratie.model.Bestelling;
import com.vdzon.administratie.model.Factuur;
import com.vdzon.administratie.model.Gebruiker;
import com.vdzon.administratie.util.SingleAnswer;
import spark.Request;
import spark.Response;

import javax.inject.Inject;

public class BestellingService {

    @Inject
    UserCrud crudService;

    protected Object putBestelling(Request req, Response res) throws Exception {
        try {
            String uuid = SessionHelper.getAuthenticatedUserUuid(req);
            Gebruiker gebruiker = crudService.getGebruiker(uuid);
            if (gebruiker == null) {
                res.status(404);
                return new SingleAnswer("not found");
            }
            String bestellingJson = req.body();
            ObjectMapper mapper = new ObjectMapper();
            BestellingDto bestellingDto = mapper.readValue(bestellingJson, BestellingDto.class);
            Bestelling bestelling = bestellingDto.toBestelling();

            removeBestelling(gebruiker, bestelling.getUuid());
            addBestelling(gebruiker, bestelling);

            crudService.updateGebruiker(gebruiker);
        } catch (Exception ex) {
            ex.printStackTrace();
            throw ex;
        }
        return new SingleAnswer("ok");
    }

    private void addBestelling(Gebruiker gebruiker, Bestelling bestelling) {
        String niewFactuurNummer = bestelling.getGekoppeldFactuurNummer();
        Factuur gekoppelFactuur = gebruiker.getDefaultAdministratie().getFactuurByFactuurNummer(niewFactuurNummer);
        if (gekoppelFactuur!=null){
            Factuur updatedFactuur = gekoppelFactuur.toBuilder().gekoppeldeBestellingNummer(bestelling.getBestellingNummer()).build();
            gebruiker.getDefaultAdministratie().removeFactuur(gekoppelFactuur.getUuid());
            gebruiker.getDefaultAdministratie().addFactuur(updatedFactuur);
        }
        gebruiker.getDefaultAdministratie().addBestelling(bestelling);
    }

    protected Object removeBestelling(Request req, Response res) throws Exception {
        try {
            String uuid = SessionHelper.getAuthenticatedUserUuid(req);
            Gebruiker gebruiker = crudService.getGebruiker(uuid);
            if (gebruiker == null) {
                res.status(404);
                return new SingleAnswer("not found");
            }
            String bestellingUuid = req.params(":uuid");
            if ("undefined".equals(bestellingUuid)) {
                bestellingUuid = null;
            }
            removeBestelling(gebruiker, bestellingUuid);
            crudService.updateGebruiker(gebruiker);
            return new SingleAnswer("ok");
        } catch (Exception ex) {
            ex.printStackTrace();
            throw ex;
        }

    }

    private void removeBestelling(Gebruiker gebruiker, String uuid) {
        Bestelling bestelling = gebruiker.getDefaultAdministratie().getBestelling(uuid);
        String factuurNummerOudeBestelling = bestelling==null ? null : bestelling.getGekoppeldFactuurNummer();
        Factuur gekoppelFactuur = gebruiker.getDefaultAdministratie().getFactuurByFactuurNummer(factuurNummerOudeBestelling);
        if (gekoppelFactuur!=null){
            Factuur updatedFactuur = gekoppelFactuur.toBuilder().gekoppeldeBestellingNummer(null).build();
            gebruiker.getDefaultAdministratie().removeFactuur(gekoppelFactuur.getUuid());
            gebruiker.getDefaultAdministratie().addFactuur(updatedFactuur);
        }
        gebruiker.getDefaultAdministratie().removeBestelling(uuid);
    }

}

File: rest/impl/src/main/java/com/vdzon/administratie/rest/bestelling/BestellingService.kt
package com.vdzon.administratie.rest.bestelling

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.vdzon.administratie.authenticatie.AuthenticationService
import com.vdzon.administratie.database.UserDao
import com.vdzon.administratie.dto.BestellingDto
import com.vdzon.administratie.model.Bestelling
import com.vdzon.administratie.model.Gebruiker
import com.vdzon.administratie.util.SingleAnswer
import spark.Request
import spark.Response
import javax.inject.Inject

class BestellingService {

    @Inject
    lateinit internal var daoService: UserDao

    @Inject
    lateinit internal var athenticationService: AuthenticationService


    @Throws(Exception::class)
    fun putBestelling(req: Request, res: Response): Any {
        val gebruiker = athenticationService.getGebruikerOrThowForbiddenException(req, res)
        val bestellingJson = req.body()
        val mapper = jacksonObjectMapper()
        val bestellingDto = mapper.readValue(bestellingJson, BestellingDto::class.java)
        val bestelling = bestellingDto.toBestelling()

        removeBestelling(gebruiker, bestelling.uuid)
        addBestelling(gebruiker, bestelling)

        daoService!!.updateGebruiker(gebruiker)
        return SingleAnswer("ok")
    }

    private fun addBestelling(gebruiker: Gebruiker, bestelling: Bestelling) {
        val niewFactuurNummer = bestelling.gekoppeldFactuurNummer
        val gekoppelFactuur = gebruiker.defaultAdministratie.getFactuurByFactuurNummer(niewFactuurNummer)
        if (gekoppelFactuur != null) {
            val updatedFactuur = gekoppelFactuur.copy(gekoppeldeBestellingNummer = bestelling.bestellingNummer)
            gebruiker.defaultAdministratie.removeFactuur(gekoppelFactuur.uuid)
            gebruiker.defaultAdministratie.addFactuur(updatedFactuur!!)
        }
        gebruiker.defaultAdministratie.addBestelling(bestelling)
    }

    @Throws(Exception::class)
    fun removeBestelling(req: Request, res: Response): Any {
        val gebruiker = athenticationService.getGebruikerOrThowForbiddenException(req, res)
        var bestellingUuid: String? = req.params(":uuid")
        if ("undefined" == bestellingUuid) {
            bestellingUuid = null
        }
        removeBestelling(gebruiker, bestellingUuid)
        daoService!!.updateGebruiker(gebruiker)
        return SingleAnswer("ok")
    }

    private fun removeBestelling(gebruiker: Gebruiker, uuid: String?) {
        val bestelling = gebruiker.defaultAdministratie.getBestelling(uuid)
        val factuurNummerOudeBestelling = bestelling?.gekoppeldFactuurNummer
        val gekoppelFactuur = gebruiker.defaultAdministratie.getFactuurByFactuurNummer(factuurNummerOudeBestelling!!)
        if (gekoppelFactuur != null) {
            val updatedFactuur = gekoppelFactuur.copy(gekoppeldeBestellingNummer = null)
            gebruiker.defaultAdministratie.removeFactuur(gekoppelFactuur.uuid)
            gebruiker.defaultAdministratie.addFactuur(updatedFactuur!!)
        }
        gebruiker.defaultAdministratie.removeBestelling(uuid)
    }

}
